File: go/day12/day12.go
package day12

import (
	"sort"
	"strings"
)

type Pos struct {
	i, j int
}

type Fence struct {
	i, j  int
	align int // 0: right, 1: down, 2: left, 3: up
}

type FenceType struct {
	coord int
	align int
}

func (f Fence) getType() (FenceType, int) {
	if f.align == 0 { // right
		return FenceType{f.j, 0}, f.i
	} else if f.align == 1 { // down
		return FenceType{f.i, 1}, f.j
	} else if f.align == 2 { // left
		return FenceType{f.j, 2}, f.i
	} else { // up
		return FenceType{f.i, 3}, f.j
	}
}

func Run(input string) (int, int) {
	// 	input = `RRRRIICCFF
	// RRRRIICCCF
	// VVRRRCCFFF
	// VVRCCCJFFF
	// VVVVCJJCFE
	// VVIVCCJJEE
	// VVIIICJJEE
	// MIIIIIJJEE
	// MIIISIJEEE
	// MMMISSJEEE`

	// 	input = `AAAAAA
	// AAABBA
	// AAABBA
	// ABBAAA
	// ABBAAA
	// AAAAAA`

	garden := make([][]rune, 0)
	for _, line := range strings.Fields(input) {
		garden = append(garden, []rune(line))
	}
	m := len(garden)
	n := len(garden[0])

	visited := make([][]bool, m)
	for i := 0; i < m; i++ {
		visited[i] = make([]bool, n)
	}

	var dfs func(i, j int) (int, int, []Fence)
	dfs = func(i, j int) (int, int, []Fence) {
		dirns := []Pos{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
		visited[i][j] = true

		fences := []Fence{}
		sides := 4
		count := 1

		createFence := func(i, j int, d Pos) {
			if d == dirns[0] {
				fences = append(fences, Fence{i, j, 0})
			} else if d == dirns[1] {
				fences = append(fences, Fence{i, j, 1})
			} else if d == dirns[2] {
				fences = append(fences, Fence{i, j, 2})
			} else if d == dirns[3] {
				fences = append(fences, Fence{i, j, 3})
			}
		}

		for _, d := range dirns {
			if i+d.i >= 0 && i+d.i < m && j+d.j >= 0 && j+d.j < n && garden[i+d.i][j+d.j] == garden[i][j] {
				if !visited[i+d.i][j+d.j] {
					a, p, f := dfs(i+d.i, j+d.j)
					sides += p
					count += a
					fences = append(fences, f...)
				}
				sides--
			} else {
				createFence(i, j, d)
			}
		}

		return count, sides, fences
	}

	sum := 0
	sum2 := 0

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if !visited[i][j] {
				a, p, fences := dfs(i, j)
				sum += a * p

				seen := map[FenceType]map[int]bool{}
				count := 0

				for _, f := range fences {
					ft, c := f.getType()

					if seen[ft] == nil {
						seen[ft] = map[int]bool{}
						seen[ft][c] = true
					} else {
						if seen[ft][c-1] || seen[ft][c+1] {
							// part of same fence
							seen[ft][c] = true
						} else {
							seen[ft][c] = true
						}
					}
				}

				for _, v := range seen {
					tcs := []int{}
					for tc, _ := range v {
						tcs = append(tcs, tc)
					}
					sort.Slice(tcs, func(i2, j2 int) bool {
						return tcs[i2] < tcs[j2]
					})
					prev := -100
					for _, tc := range tcs {
						if tc > prev+1 {
							count++
						}
						prev = tc
					}

				}

				sum2 += count * a

			}
		}
	}

	return sum, sum2
}

File: rust/src/day12.rs
use itertools::Itertools;
use std::collections::HashSet;

#[derive(PartialEq, Eq, Copy, Clone, Hash, Debug, Ord, PartialOrd)]
enum Direction {
    TOP,
    RIGHT,
    BOTTOM,
    LEFT,
}

#[derive(Eq, Hash, PartialEq, Debug)]
struct Fence {
    p: i32, // primary coordinate
    t: i32, // transverse coordinate
    dir: Direction,
}

impl Direction {
    fn get_offset(&self) -> (i32, i32) {
        match self {
            Direction::TOP => (-1, 0),
            Direction::RIGHT => (0, 1),
            Direction::BOTTOM => (1, 0),
            Direction::LEFT => (0, -1),
        }
    }
}

fn dfs(
    grid: &Vec<Vec<char>>,
    visited: &mut Vec<Vec<bool>>,
    fences: &mut HashSet<Fence>,
    i: i32,
    j: i32,
) -> (i32, i32) {
    let dirns = vec![
        Direction::TOP,
        Direction::RIGHT,
        Direction::BOTTOM,
        Direction::LEFT,
    ];
    visited[i as usize][j as usize] = true;
    let mut area = 1;
    let mut perimeter = 4;

    for dir in dirns {
        let (di, dj) = dir.get_offset();
        let ni = i + di;
        let nj = j + dj;

        if ni >= 0
            && ni < grid.len() as i32
            && nj >= 0
            && nj < grid[0].len() as i32
            && grid[ni as usize][nj as usize] == grid[i as usize][j as usize]
        {
            if !visited[ni as usize][nj as usize] {
                let (a, p) = dfs(grid, visited, fences, ni, nj);
                area += a;
                perimeter += p;
            }
            perimeter -= 1;
        } else {
            if dir == Direction::TOP || dir == Direction::BOTTOM {
                // horizontal fence
                fences.insert(Fence { p: i, t: j, dir });
            }
            if dir == Direction::LEFT || dir == Direction::RIGHT {
                // vertical fence
                fences.insert(Fence { t: i, p: j, dir });
            }
        }
    }
    (area, perimeter)
}

pub fn run(input: &str) -> (i32, i32) {
    //     let input = "RRRRIICCFF
    // RRRRIICCCF
    // VVRRRCCFFF
    // VVRCCCJFFF
    // VVVVCJJCFE
    // VVIVCCJJEE
    // VVIIICJJEE
    // MIIIIIJJEE
    // MIIISIJEEE
    // MMMISSJEEE";
    let grid = input
        .lines()
        .map(|l| l.chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();
    let m = grid.len();
    let n = grid[0].len();

    let mut visited = grid
        .iter()
        .map(|r| r.iter().map(|_| false).collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let mut cost1 = 0;
    let mut cost2 = 0;
    for i in 0..m {
        for j in 0..n {
            if !visited[i][j] {
                let mut fences = HashSet::new();
                let (a, p) = dfs(&grid, &mut visited, &mut fences, i as i32, j as i32);
                cost1 += a * p;
                let e = fences
                    .iter()
                    .sorted_by(|a, b| a.dir.cmp(&b.dir).then(a.p.cmp(&b.p)).then(a.t.cmp(&b.t)))
                    .group_by(|a| (a.dir, a.p))
                    .into_iter()
                    .map(|(_, chunk)| {
                        // println!("{:?}", chunk);
                        let mut arr = chunk.map(|c| c.t).collect::<Vec<_>>();
                        arr.sort();
                        arr.windows(2).fold(
                            1,
                            |acc, w| {
                                if w[1] - w[0] == 1 {
                                    acc
                                } else {
                                    acc + 1
                                }
                            },
                        )
                    })
                    .sum::<i32>();
                // println!("{:?}", fences);
                // println!("{:?} {:?}", a, e);
                cost2 += a * e
            }
        }
    }

    return (cost1, cost2);
}
